### PR TITLE
feat: add responsive layouts

### PR DESCRIPTION
**Pull request by HayesGordon**
This exposes `Fit.Layout` and `layoutScaleFactor` for Rive React Native.

iOS handles `layoutScaleFactor` with a value of "-1" to be not set (null). In which case iOS handles the scale factor automatically.

Android does the same, but when `layoutScaleFactor` is null. Android is more limited in it's pixel density calculations, as we cannot automatically subscribe to changes.

That means in React Native land, users will need to manage the density themselves if they want it to respond to device density changes consistently across Android and iOS - because iOS manages it better and we can't do more for Android. This will need to be a documentation effort

The current implementation attempts to handle both situations, where the value can either be null or -1.

---------
EDIT: Resolved the below!

However, on React Native we have the opposite problem. Where Android can't take a null but iOS can through the bindings 🫠 
- https://reactnative.dev/docs/legacy/native-modules-android#argument-types
- https://reactnative.dev/docs/legacy/native-modules-ios#argument-types

I'm leaning towards updating the JS to default send -1 if the prop is null. But interested to hear your thoughts, or maybe I missed some way to handle this.

### CODE REVIEW

File: example/ios/RiveReactNativeExample.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -104,6 +106,8 @@
 		ED297162215061F000B7C4FE /* JavaScriptCore.framework */ = {isa = PBXFileReference; lastKnownFileType = wrapper.framework; name = JavaScriptCore.framework; path = System/Library/Frameworks/JavaScriptCore.framework; sourceTree = SDKROOT; };
 		ED2971642150620600B7C4FE /* JavaScriptCore.framework */ = {isa = PBXFileReference; lastKnownFileType = wrapper.framework; name = JavaScriptCore.framework; path = Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS12.0.sdk/System/Library/Frameworks/JavaScriptCore.framework; sourceTree = DEVELOPER_DIR; };
 		F8AA4CA32C0F3FDB00C1A5FF /* runtime_nested_inputs.riv */ = {isa = PBXFileReference; lastKnownFileType = file; path = runtime_nested_inputs.riv; sourceTree = "<group>"; };
+		F8E2789D2CDCD6A200FAA8EF /* layouts_demo.riv */ = {isa = PBXFileReference; lastKnownFileType = file; path = layouts_demo.riv; sourceTree = "<group>"; };
+		F8E278A12CDCDAFA00FAA8EF /* layout_test.riv */ = {isa = PBXFileReference; lastKnownFileType = file; name = layout_test.riv; path = ../../../../../../rive/rive/packages/runtime_wasm/js/examples/_frameworks/layout_example/assets/layout_test.riv; sourceTree = "<group>"; };
```
Comment (lancesnider) [inline example/ios/RiveReactNativeExample.xcodeproj/project.pbxproj:110]: This is causing an error. It looks like it's pointing to a file on your local machine. Removing it from Resources in XCode fixes the issue.

**Comment by lancesnider** [comment]
This isn't a bug, but it might look like a bug to a user. On the  example if you select `Fit.Layout`, it looks like this: 

Screenshot 2024-11-07 at 9 06 16 AM

This is because this is an old file that doesn't actually have a layout. Maybe we should add one? If you have the .rev, I can do that.

**Comment by HayesGordon** [comment]
Nice 

> This isn't a bug, but it might look like a bug to a user. On the example if you select `Fit.Layout`, it looks like this:
> 
> Screenshot 2024-11-07 at 9 06 16 AM
> 
> This is because this is an old file that doesn't actually have a layout. Maybe we should add one? If you have the .rev, I can do that.

We can swap that one out for any file, it doesn't need to be the current graphic.

 I'm going to merge this PR so that we can get the feature out. If you have a file to replace, feel free to make a PR @lancesnider 🙏

**Comment by HayesGordon** [comment]
> This isn't a bug, but it might look like a bug to a user. On the example if you select `Fit.Layout`, it looks like this:
> 
> Screenshot 2024-11-07 at 9 06 16 AM
> 
> This is because this is an old file that doesn't actually have a layout. Maybe we should add one? If you have the .rev, I can do that.

Nice catch. I don't have the .rev, but we can use any example! If you have one to commit to this PR then let's do, otherwise, let's get this feature out and update the sample app later.

File: example/ios/RiveReactNativeExample.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -724,7 +734,7 @@
 					"-ObjC",
 					"-lc++",
 				);
-				PRODUCT_BUNDLE_IDENTIFIER = com.example.rivereactnative;
+				PRODUCT_BUNDLE_IDENTIFIER = gordon.example.rivereactnative;
```
Comment (dskuza) [inline example/ios/RiveReactNativeExample.xcodeproj/project.pbxproj:737]: Probably fine, just calling this out as a thing that's been committed.

File: ios/RiveReactNativeView.swift [hunk 3]
```diff
@@ -41,6 +41,8 @@ class RiveReactNativeView: RCTView, RivePlayerDelegate, RiveStateMachineDelegate
     
     @objc var fit: String?
     
+    @objc var layoutScaleFactor: NSNumber = -1.0 // -1.0 will inform the iOS runtime to determine the correct scale factor automatically
```
Comment (dskuza) [inline ios/RiveReactNativeView.swift:44]: This type is a `Double` in Swift (and likely ObjC by interop). What's the decision behind using NSNumber?

File: ios/RiveReactNativeView.swift [hunk 4]
```diff
@@ -99,6 +101,10 @@ class RiveReactNativeView: RCTView, RivePlayerDelegate, RiveStateMachineDelegate
         if (changedProps.contains("alignment"))  {
             viewModel?.alignment = convertAlignment(alignment)
         }
+        
+        if (changedProps.contains("layoutScaleFactor"))  {
+            viewModel?.layoutScaleFactor = layoutScaleFactor.doubleValue
```
Comment (dskuza) [inline ios/RiveReactNativeView.swift:106]: Ah, usage is here as a double value. But, still curious!

File: ios/RiveReactNativeViewManager.m [hunk 5]
```diff
@@ -6,6 +6,7 @@ @interface RCT_EXTERN_MODULE(RiveReactNativeViewManager, RCTViewManager)
 RCT_EXPORT_VIEW_PROPERTY(resourceName, NSString)
 RCT_EXPORT_VIEW_PROPERTY(url, NSString)
 RCT_EXPORT_VIEW_PROPERTY(fit, NSString)
+RCT_EXPORT_VIEW_PROPERTY(layoutScaleFactor, NSNumber)
```
Comment (dskuza) [inline ios/RiveReactNativeViewManager.m:9]: I'm assuming this has something to do with using NSNumber over Double?

**Review by dskuza — APPROVED** [review]
This looks good to me! Just pointing one small thing out, and a question about another (specific to iOS).

**Review by lancesnider — APPROVED** [review]

File: example/ios/RiveReactNativeExample.xcodeproj/project.pbxproj (hunk 2 above)
Comment (HayesGordon) [reply example/ios/RiveReactNativeExample.xcodeproj/project.pbxproj:737]: Ah I forgot this, will remove

File: ios/RiveReactNativeView.swift (hunk 3 above)
Comment (HayesGordon) [reply ios/RiveReactNativeView.swift:44]: Double doesn't work for nullable(optional) values in these bindings: https://reactnative.dev/docs/legacy/native-modules-ios#argument-types

File: ios/RiveReactNativeViewManager.m (hunk 5 above)
Comment (HayesGordon) [reply ios/RiveReactNativeViewManager.m:9]: Yup! In one of the other comments I linked to the reference docs

**Pull request merged**